### PR TITLE
[DRAFT] Support raw inputs in CTC Loss Python API

### LINKED ISSUE #52241: CTCLoss gradient is incorrect
## 🐛 Bug
Hi,

While working on some CTC extensions, I noticed that torch's CTCLoss was computing incorrect gradient. At least when using CPU (I have not tested on GPU yet). I observed this problem on both pytorch 1.3 and pytorch 1.7.1 , using MacOS/anaconda distribution.

## To Reproduce
Torch's own gradcheck tool detect this problem. Here is a minimal example of gradient computation going wrong (for the record, it is not a problem of precision with gradcheck. The computed gradient is really wrong.

```python
import torch
import torch.nn as nn

# For this minimal example, let us use a mninbatch of size 1, 3 time steps and a label vocabulary size of 2
V = 2
T = 3

# Generate some random input parameters for CTC
params_ctc = nn.functional.log_softmax(torch.randn(T, 1, V+1)) 
params_ctc = params_ctc.double()  # convert to double so that gradcheck is more accurate
params_ctc.requires_grad = True

torch_ctc = nn.CTCLoss() 

labels = torch.tensor([[1]] )
labels_size = torch.tensor([1])
T_lengths = torch.tensor([T])

torch.autograd.gradcheck(torch_ctc, [params_ctc, 
        labels ,T_lengths, labels_size])

```

## Expected behavior

The gradcheck call above should pass without error.

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): MacOS
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version:   N/A (CPU)
 - GPU models and configuration: N/A
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by redwrasse**
(Maybe) Fixes #52241 

Adds support for unnormalized inputs to the Python API(s) for CTC loss.

Draft PR: would appreciate permission for a CI run to check tests pass (cc @soulitzer) (CTC loss OpInfo and forward pass tests pass locally) and feedback on this approach from a maintainer and original implementation author (cc @soulitzer and @t-vi, this follows up on your comment https://github.com/pytorch/pytorch/issues/52241#issuecomment-778724613) on this approach. 

This should be backwards-compatible- as log_softmax is idempotent, behavior on the currently supported log-simplex space is preserved, and simply extended to include raw inputs. (See https://github.com/pytorch/pytorch/issues/52241#issuecomment-2973324675 for reference.)

### CODE REVIEW

**Comment by albertz** [comment]
But that somehow misses the point of #52241, or at least one important aspect. There is the valid use case that people want to use unnormalized scores directly, and log_softmax should not be applied internally as done here!

I think the obvious correct solution is to simply fix the gradient of ctc_loss. There are a couple of workarounds proposed in #52241 which do that, for example https://github.com/pytorch/pytorch/issues/52241#issuecomment-2502129316, but it would be much cleaner if PyTorch would directly provide the correct gradient.

**Comment by redwrasse** [comment]
> But that somehow misses the point of #52241, or at least one important aspect. There is the valid use case that people want to use unnormalized scores directly, and log_softmax should not be applied internally as done here!
> 
> I think the obvious correct solution is to simply fix the gradient of ctc_loss. There are a couple of workarounds proposed in #52241 which do that, for example [#52241 (comment)](https://github.com/pytorch/pytorch/issues/52241#issuecomment-2502129316), but it would be much cleaner if PyTorch would directly provide the correct gradient.

@albertz could you clarify your point? A code snippet indicating the desired numerical behavior would help.

My working assumptions based on a not thorough reading of #52241 are:

i) from https://github.com/pytorch/pytorch/issues/52241#issuecomment-1169951867 the cuDNN implementation encapsulates a log_softmax, so we're stuck with it; since it's idempotent, might as well add another one globally.
ii) changes should ideally be backwards-compatible

And I'm proposing these are met by redefining the operation to be

```torch_ctc = lambda x, *o: torch.nn.functional.ctc_loss(torch.nn.functional.log_softmax(x, -1), *o)```

**Comment by albertz** [comment]
> i) from [#52241 (comment)](https://github.com/pytorch/pytorch/issues/52241#issuecomment-1169951867) the cuDNN implementation encapsulates a log_softmax, so we're stuck with it; since it's idempotent, might as well add another one globally.

No. The forward pass of ctc_loss does not encapsulate log_softmax. It just calculates the loss as it is defined, i.e. as $-\log \sum_{y[0:T] : \textrm{targets}} \prod_t \exp (\textrm{logprobs}[t, y_t])$. This is what is being correctly computed. This can also be computed when logprobs is not normalized. And this is a valid use case that some people want (e.g. sometimes you want to include a prior probability or so).

However, the gradient is incorrect. The grad of ctc_loss (let's call it grad_ctc_loss) does not correctly calculate the gradient for this calculation. The gradient is as-if it would be with an additional log_softmax.

However, the current incorrect gradient is easy to transform into the correct gradient, which is what is done in some of the suggestions there, for example https://github.com/pytorch/pytorch/issues/52241#issuecomment-2502129316.

> ii) changes should ideally be backwards-compatible

We should differentiate maybe two use cases:

a) The user anyway already did log_softmax before. (And also, this is the only use of the log_probs in the computation graph.)
b) The user did not use log_softmax but sth else, and this did not necessarily result in a normalized prob distrib, and this is exactly intended that it is not normalized.

For a), there was no problem, and there is no problem by any potential fix. Your change does not change anything for this. And fixing the grad_ctc_loss would also not change anything.
For b), your change would break this use case. This use case would not be possible at all anymore. However, fixing the grad_ctc_loss that it calculates the correct gradient just means that users need to deactivate the workaround which is currently used.

So, I would argue, just fixing grad_ctc_loss makes more sense.

**Comment by redwrasse** [comment]
@albertz thanks, yes I agree with your points. I wasn't aware the current forward-pass is correct for unnormalized inputs and doesn't absorb a log-softmax. And yes agree the way forward is to fix the gradient to be wrt. unnormalized inputs. I'll close this PR.
